pkg/controller/controller.go

```diff
@@ -74,7 +75,7 @@ type Controller interface {
 	// Watch may be provided one or more Predicates to filter events before
 	// they are given to the EventHandler.  Events will be passed to the
 	// EventHandler if all provided Predicates evaluate to true.
-	Watch(src source.Source, eventhandler handler.EventHandler, predicates ...predicate.Predicate) error
+	Watch(cluster cluster.Cluster, src source.Source, eventhandler handler.EventHandler, predicates ...predicate.Predicate) error
```

Comment:
This seems like a stopgap. The controller doesn't need this at all, nor do all source.Source and handler.EventHandler implementations. I'd rather unexport the ones that need it and export a new constructor that takes a cluster.Cluster to end up with one breaking change only

Reply:
That's true, the controller doesn't, but the Source and EventHandler need the fields on the Cluster/Manager to set their caches, and so on. Do you have a better alternative in mind?

Removing the dependency injection from the Source is going to probably take a major effort, which not sure if we should tackle just yet, but open to better alternatives. The tricky part is going to be later on to make sure Sources, specifically the source.Kind one, and other most used ones, can be used w/ specific clusters.

Reply:
Why do you think it will be a major effort? There is exactly one source that uses this, `source.Kind`. We can unexport the type and make `Kind` a func that matches what we have in `NewKindWithCache`. 

Same story for the handler, the only one that actually needs this is `EnqueueRequestForOwner`. Lets unexport the type and instead make `*EnqueueRequestForOwner` a constructor func. Other than that there is only `*enqueueRequestsFromMapFunc` which just passes on `Inject.Funcs` but doens't use that itself.


> The tricky part is going to be later on to make sure Sources, specifically the source.Kind one, and other most used ones, can be used w/ specific clusters.

This is already possible today, using `source.NewKindWithCache`.